File: java/src/main/java/vsdk/toolkit/common/dataStructures/CircularListBase.java
package vsdk.toolkit.common.dataStructures;

public class CircularListBase {
    private CircularListLink last;

    public CircularListBase() {
        last = null;
    }

    public void addLink(CircularListLink data) {
        if (last != null) {
            data.nextLink = last.nextLink;
        }
        else {
            last = data;
        }
        last.nextLink = data;
    }

    public void appendLink(CircularListLink data) {
        if (last != null) {
            data.nextLink = last.nextLink;
            last = last.nextLink = data;
        }
        else {
            last = data;
            data.nextLink = data;
        }
    }

    public CircularListLink remove() {
        if (last == null) {
            return null;
        }

        CircularListLink first = last.nextLink;
        if (first == last) {
            last = null;
        }
        else {
            last.nextLink = first.nextLink;
        }

        return first;
    }

    public void clear() {
        last = null;
    }

    public CircularListLink lastLink() {
        return last;
    }
}

File: typescript/src/main/vsdk/toolkit/common/dataStructures/CircularListBase.ts
import { CircularListLink } from "./CircularListLink";

export class CircularListBase {
  private last: CircularListLink | null;

  public constructor() {
    this.last = null;
  }

  public addLink(data: CircularListLink): void {
    if (this.last !== null) {
      data.nextLink = this.last.nextLink;
    }
    else {
      this.last = data;
    }
    this.last.nextLink = data;
  }

  public appendLink(data: CircularListLink): void {
    if (this.last !== null) {
      data.nextLink = this.last.nextLink;
      this.last = this.last.nextLink = data;
    }
    else {
      this.last = data;
      data.nextLink = data;
    }
  }

  public remove(): CircularListLink | null {
    if (this.last === null) {
      return null;
    }

    const first = this.last.nextLink as CircularListLink;
    if (first === this.last) {
      this.last = null;
    }
    else {
      this.last.nextLink = first.nextLink;
    }

    return first;
  }

  public clear(): void {
    this.last = null;
  }

  public lastLink(): CircularListLink | null {
    return this.last;
  }
}
